include/dxc/DXIL/DxilFunctionProps.h

```diff
@@ -25,6 +25,93 @@ class Constant;
 } // namespace llvm
 
 namespace hlsl {
+
+// SM 6.6 allows WaveSize specification for only a single required size.
+// SM 6.8+ allows specification of WaveSize as a min, max and preferred value.
+struct DxilWaveSize {
+  unsigned Min = 0;
+  unsigned Max = 0;
+  unsigned Preferred = 0;
+
+  DxilWaveSize() = default;
+  DxilWaveSize(unsigned min, unsigned max = 0, unsigned preferred = 0)
+      : Min(min), Max(max), Preferred(preferred) {}
+  DxilWaveSize(const DxilWaveSize &other) = default;
+  DxilWaveSize &operator=(const DxilWaveSize &other) = default;
+  bool operator==(const DxilWaveSize &other) const {
+    return Min == other.Min && Max == other.Max && Preferred == other.Preferred;
+  }
+
+  // Encode valid DxilWaveSize, handling potential degenerate cases.
+  static DxilWaveSize Encode(unsigned min, unsigned max = 0,
+                             unsigned preferred = 0) {
+    if (max == min)
+      max = 0;
+    if (max == 0 && preferred == min)
+      preferred = 0;
+    return DxilWaveSize(min, max, preferred);
+  }
+
+  // Valid non-zero values are powers of 2 between 4 and 128, inclusive.
+  static bool IsValidValue(unsigned Value) {
+    return (Value >= 4 && Value <= 128 && ((Value & (Value - 1)) == 0));
+  }
+  // Valid encodings:
```

Comment:
This comment makes it sound like these _encodings_ are valid. IIUC the old wave size didn't encode anything if the values were `0`, I assume we should do the same here.

There is also an outstanding question for the spec about whether or not we're deprecating the old metadata tag or not. I think that impacts which other formulations are valid. I filed this issue to track that:
https://github.com/microsoft/hlsl-specs/issues/160

Reply:
"encodings" is referring to the encoding of different forms *in this structure*.  What would you prefer?

Regarding deprecation of the old metadata, The prior change already made that decision (not to deprecate it), and I already had an issue filed for that question: microsoft/hlsl-specs#158.

We can change the decision, but this PR doesn't change the current state.  Besides, these "encodings" still only pertain to the encoding of various valid states into this structure, so that issue doesn't impact this code - only how to write these states into the metadata in `DxilMetadataHelper.cpp`.

Reply:
valid "values" or "representations" would be more clear. "Encodings" implies that it is being encoded, not that the values are just stored.

Reply:
When writing the spec, I had no intention of deprecating the old tag. In fact, the spec says that the only valid use of the new tag is with at least two values, which this deviates from. I don't really care enough to insist, but that's what is currently specced.

Reply:
Okay. reread my own words and I see that I didn't actually write that. It's what I was envisioning, but I'm not insistent. I should clarify the spec though.

Reply:
I can replace the word "encodings" in the comment with "representations" if that will satisfy.  I will also have to rename two methods, but now I don't have a good name for them (encode/decode naming no longer works).

All the different states for WaveSize must be "encoded" into the values of this structure in a specific way for it to be valid, including the state where it is undefined.  You could also say they must be "represented" in the values of this structure in a specific way for it to be valid.  It seems to be a minor semantic point.

Also: there are multiple "encodings": The internal representation in this structure, one in llvm metadata, one in PSV0, and one in RDAT.  This does not describe any "encodings" besides the one in this structure.

The spec issue is **not** something that impacts this code, or the comment here.  This isn't referring to just the "encoding" specifically into metadata.  Note: "encodings" for PSV0 and RDAT are different than the one for metadata as well.